### PR TITLE
[Cairo] Initial integration

### PR DESCRIPTION
**Pull request by ecalp-tps**
Hello,

I am a Software Engineering Intern at Google, working with the Third Party Security (TPS) team. As part of my internship I am working on fuzzing GNOME/freedesktop libraries and Cairo is one of the libraries that we have picked for this work.

You can find the MR for integrating my fuzzers into the main Cairo repository [here](https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/69). I've been in touch with the upstream about the OSS-Fuzz integration and their main concern is the lack of development staff on that end. For now I’m fetching the Cairo repo from my personal fork in the Dockerfile. I'm sending in the PR since my internship ends this week.

Thanks!
Ceyhun
ceyhunalp@google.com

### CODE REVIEW

**Comment by inferno-chromium** [comment]
We would prefer if you fuzz from trunk branch and add a patch.diff with anything needed. usually you can start with build.sh and fuzzers in this repo itself (/CL).

Are they ready to accept bugs ? that is sufficient enough, ask them to give an email id.

**Review by inferno-chromium — CHANGES_REQUESTED** [review]
See question above.

**Comment by inferno-chromium** [comment]
Also, please fix / disable memory and undefined builds by specifying in project.yaml sanitizers: field.

File: projects/cairo/project.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+homepage: https://gitlab.freedesktop.org/cairo/cairo
+language: c
+primary_contact: security-tps@google.com
```
Comment (inferno-chromium) [inline projects/cairo/project.yaml:3]: Try getting actual developer email. I dont think this mailing list is correct to add him, check with Julien, Jose. @Google-Autofuzz

**Review by inferno-chromium — APPROVED** [review]
Can you ask Julien to review this PR as well.

**Comment by inferno-chromium** [comment]
For ubsan, check if this fixes it in build.sh, these checks should be later enabled, so add a TODO.

export CFLAGS="$CFLAGS -fno-sanitize=vptr"
export CXXFLAGS="$CXXFLAGS -fno-sanitize=vptr"

File: projects/cairo/project.yaml (hunk 1 above)
Comment (ecalp-tps) [reply projects/cairo/project.yaml:3]: I contacted the upstream yesterday based on your comment: https://gitlab.freedesktop.org/cairo/cairo/-/merge_requests/69#note_705957

I'll update this as soon as I hear back from them.

File: projects/cairo/targets/pdf_surface_fuzzer.c [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    if (status != CAIRO_STATUS_SUCCESS) {
+        fuzzer_release_tmpfile(tmpfile);
+        return 0;
+    }
+
+    char *buf = (char *) malloc(size + 1);
+    memcpy(buf, data, size);
+    buf[size] = '\0';
+
+    flags = CAIRO_PDF_OUTLINE_FLAG_BOLD | CAIRO_PDF_OUTLINE_FLAG_OPEN;
```
Comment (Google-Autofuzz) [inline projects/cairo/targets/pdf_surface_fuzzer.c:42]: It seems that this variable is unused?

File: projects/cairo/targets/text_glyphs_fuzzer.c [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+    cairo_t *cr;
+    cairo_surface_t *surface;
+    cairo_status_t status;
+    cairo_text_extents_t extents;
+    cairo_text_cluster_t cluster;
+
+    // Taken from test/text-glyph-range.c
+    long int index[] = {
```
Comment (Google-Autofuzz) [inline projects/cairo/targets/text_glyphs_fuzzer.c:26]: Shouldn't those be random?

**Review by Google-Autofuzz — CHANGES_REQUESTED** [review]

**Comment by inferno-chromium** [comment]
Check ubsan failures, fix them or disable it.

**Comment by inferno-chromium** [comment]
Merging for now.

File: projects/cairo/targets/pdf_surface_fuzzer.c (hunk 2 above)
Comment (ecalp-tps) [reply projects/cairo/targets/pdf_surface_fuzzer.c]: removed.

**Comment by ecalp-tps** [comment]
> Check ubsan failures, fix them or disable it.

UBSan failure is a run-time error due to a null-pointer dereference in one of the fuzzers when a clean-up function (`cairo_surface_destroy`) is called. I talked to Julien and Michael and we think that this might be due to an issue in Cairo so I didn't disable UBSan.

**Pull request merged**